Extension/src/LanguageServer/copilotProviders.ts

```diff
@@ -0,0 +1,136 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+'use strict';
+
+import * as vscode from 'vscode';
+import * as util from '../common';
+import * as telemetry from '../telemetry';
+import { ChatContextResult, GetIncludesResult } from './client';
+import { getActiveClient } from './extension';
+
+let isRelatedFilesApiEnabled: boolean | undefined;
+
+export interface CopilotTrait {
+    name: string;
+    value: string;
+    includeInPrompt?: boolean;
+    promptTextOverride?: string;
+}
+
+export interface CopilotApi {
+    registerRelatedFilesProvider(
+        providerId: { extensionId: string; languageId: string },
+        callback: (
+            uri: vscode.Uri,
+            context: { flags: Record<string, unknown> },
+            cancellationToken: vscode.CancellationToken
+        ) => Promise<{ entries: vscode.Uri[]; traits?: CopilotTrait[] }>
+    ): Disposable;
+}
+
+export async function registerRelatedFilesProvider(): Promise<void> {
+    if (!await getIsRelatedFilesApiEnabled()) {
+        return;
+    }
+
+    const api = await getCopilotApi();
+    if (util.extensionContext && api) {
+        try {
+            for (const languageId of ['c', 'cpp', 'cuda-cpp']) {
+                api.registerRelatedFilesProvider(
+                    { extensionId: util.extensionContext.extension.id, languageId },
+                    async (_uri: vscode.Uri, context: { flags: Record<string, unknown> }, token: vscode.CancellationToken) => {
+
+                        const getIncludesHandler = async () => {
+                            return (await getIncludesWithCancellation(1, token))?.includedFiles.map(file => vscode.Uri.file(file)) ?? [];
+                        }
+                        const getTraitsHandler = async () => {
+                            const chatContext: ChatContextResult | undefined = await (getActiveClient().getChatContext(token) ?? undefined);
+
+                            if (!chatContext) {
+                                return undefined;
+                            }
+
+                            let traits: CopilotTrait[] = [
+                                { name: "language", value: chatContext.language, includeInPrompt: true, promptTextOverride: `The language is ${chatContext.language}.` },
```

Comment:
Copilot already includes either a "language marker" or a "path marker" (but not both) at the top of the prompt. This particular trait may end up being redundant with the language marker, but as long as we have independent experimental control, I think it's fine to leave.

Reply:
We can control it with a string[] flag like the following, would that be ok? Thanks.
![image](https://github.com/user-attachments/assets/2ff93a3b-9b7b-42ae-aef8-5eccfaaf0854)

Reply:
I think that will work. We can't actually set an experimentation variable to `string[]` type in the ExP UI (only boolean, number, or string) but that shouldn't matter because there will be an translation layer in copilot-client anyway where we can for example convert a `string` to a `string[]`.

Reply:
I updated both my PRs, thanks.